julia_src/reopt_model.jl

```diff
@@ -1145,8 +1145,202 @@ function add_cost_function(m, p)
     #= Note: 0.9999*m[:MinChargeAdder] in Obj b/c when m[:TotalMinCharge] > (TotalEnergyCharges + m[:TotalDemandCharges] + TotalExportBenefit + m[:TotalFixedCharges])
 		it is arbitrary where the min charge ends up (eg. could be in m[:TotalDemandCharges] or m[:MinChargeAdder]).
 		0.0001*m[:MinChargeAdder] is added back into LCC when writing to results.  =#
+
+	# Add climate costs
+	if p.Include_climate_in_objective # if user selects to include climate in objective
+		add_to_expression!(Costs, m[:Lifecycle_Emissions_Cost_CO2])
+	end
+
+	# Add Health costs (NOx, SO2, PM2.5)
+	if p.Include_health_in_objective
+		add_to_expression!(Costs, m[:Lifecycle_Emissions_Cost_Health])
+	end
+
+end
+
+# Renewable electricity calculation
+function add_re_elec_calcs(m,p)
+	if p.IncludeExportedREElecinTotal
+		include_exported_re_elec_in_total = 1
+	else
+		include_exported_re_elec_in_total = 0
+	end
+	m[:AnnualREEleckWh] = @expression(m,p.TimeStepScaling*
+		(sum(p.ProductionFactor[t,ts] * p.LevelizationFactor[t] * m[:dvRatedProduction][t,ts] * p.TechPercentRE[t] for t in p.ElectricTechs, ts in p.TimeStep) #total RE elec generation, excl steam turbine
+		- sum(m[:dvProductionToStorage][b,t,ts]*p.TechPercentRE[t]*(1-p.ChargeEfficiency[t,b]*p.DischargeEfficiency[b]) for t in p.ElectricTechs, b in p.ElecStorage, ts in p.TimeStep) #minus battery efficiency losses
+		- sum(m[:dvProductionToCurtail][t,ts]*p.TechPercentRE[t] for t in p.ElectricTechs, ts in p.TimeStep) # minus curtailment.
+		- (1-include_exported_re_elec_in_total)*sum(m[:dvProductionToGrid][t,u,ts]*p.TechPercentRE[t] for t in p.ElectricTechs,  u in p.ExportTiersByTech[t], ts in p.TimeStep) # minus exported RE, if RE accounting method = 0.
+		+ p.STElecOutToThermInRatio * sum(m[:dvThermalToSteamTurbine][tst,ts]*p.TechPercentRE[tst] for ts in p.TimeStep, tst in p.TechCanSupplySteamTurbine) )) # plus steam turbine generation (steam turbine p.TechPercentRE input = 0 because it is a decision variable dependent on fraction of steam generated by RE fuel
+		# TODO: Steam Turbine battery losses and curtailment (challenge is that equations seem non-linear)
+		# Note: if battery ends up being allowed to discharge to grid, need to make sure only RE that is being consumed onsite is counted so battery doesn't become a back door for RE to grid.
```

Comment:
If the battery is grid tied, then we need a method of keeping track of the %RE and avg emissions factor of the energy that is in the battery at every time step. If we don't do this, then grid exports must be counted in the %RE calculation because RE going into the battery can't be separated into what will eventually go to grid vs load. And at the same time, grid exports can't be counted in the emissions totals because energy from battery to grid can't be separated into components based on original source emissions factor. This behavior would be inconsistent between the %RE and emissions calcs. I think the most accurate way to keep track would be to have a battery %RE expression and battery emissions factor expression, both calculated as every timestep based on the amounts of incoming energy from each source and amount of outgoing energy (would have to decide whether charge or discharge happens first). The other option is to never count anything that goes into storage. This would be technically less accurate, but also could be the way to go if it aligns with some existing emissions accounting standards.

Reply:
The battery can never export to the grid in the current model. Does that resolve this issue?

Reply:
Yes if we don't intend to let the battery discharge to grid then it's not an issue

Reply:
I think it's fine to not worry about something that's not yet in the model, but good to keep a short note in there for if/when we do add it.